File: deployment.py
from http.server import BaseHTTPRequestHandler, HTTPServer
import json
import os
import subprocess
from typing import TypedDict


with open(".env", "r") as f:
    for line in f:
        if "DEPLOYMENT_TOKEN" in line:
            deployment_token = line.strip().split('=')[1]
if not deployment_token:
    print('Deployment token not found in .env file', flush=True)
    exit(1)


app2compose = {
    "backend": "compose.backend.yaml",
    "frontend": "compose.frontend.yaml",
}


class DockerImage(TypedDict):
    image: str
    compose_file: str
    env_variable: str


class DeploymentHandler(BaseHTTPRequestHandler):
    def do_POST(self):
        if '/deployment' in self.path:
            self.handle_deploy()
        else:
            self.send_response(404)
            self.end_headers()
            self.wfile.write(b'Not Found')

    def handle_deploy(self):
        param_dict = self.get_query_params()
        if param_dict.get("token") != deployment_token:
            self.send_response(401)
            self.end_headers()
            self.wfile.write(b'Unauthorized')
            return

        if param_dict.get("app") not in app2compose:
            self.send_response(400)
            self.end_headers()
            self.wfile.write(b'Invalid application')
            return

        app = param_dict["app"]
        body = json.loads(self.rfile.read(
            int(self.headers['Content-Length'])))
        image = body["repository"]["repo_name"] + \
            ":" + body["push_data"]["tag"]

        with open('docker_images.json', 'r') as f:
            images: dict[str, DockerImage] = json.load(f)
            images[app]["image"] = image
        with open('docker_images.json', 'w') as f:
            json.dump(images, f, indent=2)
        deploy_image(images[app])

        self.send_response(200)
        self.end_headers()
        self.wfile.write(b'Deployment request received')

    def get_query_params(self) -> dict:
        if '?' not in self.path:
            return {}
        query = self.path.split('?')[1]
        param_dict = {}
        for param in query.split('&'):
            key, value = param.split('=')
            if key in param_dict:
                if isinstance(param_dict[key], list):
                    param_dict[key].append(value)
                else:
                    param_dict[key] = [param_dict[key], value]
            else:
                param_dict[key] = value
        return param_dict


def run(server_class=HTTPServer, handler_class=DeploymentHandler, port=5000):
    server_address = ('', port)
    httpd = server_class(server_address, handler_class)
    print(f'Starting http server on port {port}', flush=True)
    httpd.serve_forever()


def deploy_image(image: DockerImage):
    # set image as environment variable
    os.environ[image['env_variable']] = image['image']
    try:
        subprocess.run(['docker', 'compose', '-f', image['compose_file'],
                        'down'], check=True, capture_output=True, text=True)
    except subprocess.CalledProcessError as e:
        print(
            f'Error occurred during docker-compose up: {e.stderr}', flush=True)
        return
    try:
        subprocess.run(['docker', 'compose', '-f', image['compose_file'],
                        'up', '-d'], check=True, capture_output=True, text=True)
    except subprocess.CalledProcessError as e:
        print(
            f'Error occurred during docker-compose up: {e.stderr}', flush=True)
        return
    print('Docker Compose triggered successfully', flush=True)


if __name__ == '__main__':
    run()

File: deployment.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"os/exec"
	"time"

	"github.com/joho/godotenv"
)

type WebhookRequest struct {
	CallbackUrl string `json:"callback_url"`
	PushData    struct {
		PushedAt int    `json:"pushed_at"`
		Pusher   string `json:"pusher"`
		Tag      string `json:"tag"`
	} `json:"push_data"`
	Repository struct {
		CommentCount    int    `json:"comment_count"`
		DateCreated     int    `json:"date_created"`
		Description     string `json:"description"`
		Dockerfile      string `json:"dockerfile"`
		FullDescription string `json:"full_description"`
		IsOfficial      bool   `json:"is_official"`
		IsPrivate       bool   `json:"is_private"`
		IsTrusted       bool   `json:"is_trusted"`
		Name            string `json:"name"`
		Namespace       string `json:"namespace"`
		Owner           string `json:"owner"`
		RepoName        string `json:"repo_name"`
		RepoUrl         string `json:"repo_url"`
		StarCount       int    `json:"star_count"`
		Status          string `json:"status"`
	} `json:"repository"`
}

type DockerImage struct {
	Tag        string     `json:"tag"`
	app        string     `json:"compose_file"`
	Repository string     `json:"repository"`
	Timestamp  *time.Time `json:"timestamp"`
}
type DockerImageOut struct {
	Tag        string    `json:"tag"`
	Repository string    `json:"repository"`
	Timestamp  time.Time `json:"timestamp"`
}
type DockerImages []DockerImage
type DockerImagesOut []DockerImageOut

func (images *DockerImages) MarshalJSON() ([]byte, error) {
	imagesOut := make(DockerImagesOut, 0)
	for _, image := range *images {
		if image.Timestamp != nil {
			imagesOut = append(imagesOut, DockerImageOut{
				Tag:        image.Tag,
				Repository: image.Repository,
				Timestamp:  *image.Timestamp,
			})
		}
	}
	return json.Marshal(imagesOut)
}

var images = DockerImages{
	{
		Repository: "liberatorist/bpl2-backend",
		Tag:        "",
		app:        "backend",
		Timestamp:  nil,
	},
	{
		Repository: "liberatorist/bpl2-frontend",
		Tag:        "",
		app:        "frontend",
		Timestamp:  nil,
	},
}

func postDeploymentHandler(w http.ResponseWriter, r *http.Request) {
	if r.URL.Query().Get("token") != os.Getenv("DEPLOYMENT_TOKEN") {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}
	body := WebhookRequest{}
	err := json.NewDecoder(r.Body).Decode(&body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	for i, image := range images {
		if image.Repository == body.Repository.RepoName {
			image.Tag = body.PushData.Tag
			now := time.Now()
			image.Timestamp = &now
			err := deployImage(image)
			if err != nil {
				http.Error(w, err.Error(), http.StatusInternalServerError)
				return
			}
			images[i] = image
			w.Write([]byte("Deployment successful"))
			return
		}
	}
	http.Error(w, "Unknown repository", http.StatusBadRequest)
}

func deploymentHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method == "POST" {
		postDeploymentHandler(w, r)
		return
	}
	if r.Method == "GET" {
		imagesJson, err := images.MarshalJSON()
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		w.Write(imagesJson)
		return
	}
	http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
}

func runShellCommand(command ...string) error {
	cmd := exec.Command(command[0], command[1:]...)
	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	err := cmd.Run()
	if err != nil {
		return fmt.Errorf("error occurred during %v: %v", command, stderr.String())
	}
	return nil
}

func deployImage(image DockerImage) error {
	os.Setenv("IMAGE", image.Repository+":"+image.Tag)
	err := runShellCommand("docker", "pull", image.Repository+":"+image.Tag)
	if err != nil {
		return err
	}
	err = runShellCommand("docker", "compose", "-f", image.app+"/docker-compose.yaml", "down")
	if err != nil {
		return err
	}
	err = runShellCommand("docker", "compose", "-f", image.app+"/docker-compose.yaml", "up", "-d")
	if err != nil {
		return err
	}
	return nil
}

func main() {
	err := godotenv.Load()
	if err != nil {
		log.Fatal("Error loading .env file")
	}
	http.HandleFunc("/deployment", deploymentHandler)

	err = http.ListenAndServe(":5000", nil)
	if err != nil {
		log.Fatal("ListenAndServe: ", err)
	}

}
